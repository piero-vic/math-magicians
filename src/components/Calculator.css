* {
  box-sizing: border-box;
}

.Calculator {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  width: 400px;
  border: 1px solid grey;
}

.output-bar {
  grid-column: span 4;
  text-align: right;
  padding: 20px;
  border: 1px solid grey;
  color: white;
  background: grey;
  word-wrap: break-word;
  word-break: break-all;
}

button {
  border: 1px solid grey;
  padding: 20px;
}

.btn0 {
  grid-column: span 2;
}

.operation-btn {
  background-color: orange;
}
