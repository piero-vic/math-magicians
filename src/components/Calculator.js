import React, { useState } from 'react';
import './Calculator.css';
import Button from './Button';
import calculate from '../logic/calculate';

const Calculator = () => {
  const [data, updateData] = useState({
    total: null,
    next: null,
    operation: null,
  });

  const buttons = ['AC', '+/-', '%', 'รท',
    '7', '8', '9', 'x',
    '4', '5', '6', '-',
    '1', '2', '3', '+',
    '0', '.', '='];

  return (
    <div className="Calculator">
      <div className="output-bar">
        {data.next || data.operation || data.total || '0'}
      </div>
      {
        buttons.map((btn) => (
          <Button
            key={btn.id}
            className={
              btn === 'รท' || btn === 'x' || btn === '-' || btn === '+' || btn === '='
                ? 'operation-btn'
                : `btn${btn}`
            }
            data={data}
            buttonClick={() => updateData(calculate(data, btn))}
            name={btn}
          />
        ))
      }
    </div>
  );
};

export default Calculator;
