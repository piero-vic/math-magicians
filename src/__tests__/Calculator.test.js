import React from 'react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';
import Calculator from '../components/Calculator';

it('renders correctly', () => {
  const tree = renderer
    .create(<Calculator />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('Buttons runs operations and updates the UI', () => {
  const component = render(<Calculator />);

  const outputBar = component.container.querySelector('.output-bar');
  expect(outputBar).toHaveTextContent('0');

  const btn9 = component.container.querySelector('.btn9');
  expect(btn9).toHaveTextContent('9');

  const sumButton = component.getByText('+');
  const equalButton = component.getByText('=');

  fireEvent.click(btn9);
  fireEvent.click(sumButton);
  fireEvent.click(btn9);
  fireEvent.click(equalButton);

  expect(outputBar).toHaveTextContent('18');
});
