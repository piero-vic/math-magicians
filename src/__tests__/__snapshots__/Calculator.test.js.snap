// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Calculator"
>
  <div
    className="output-bar"
  >
    0
  </div>
  <button
    className="btnAC"
    onClick={[Function]}
    type="button"
  >
    AC
  </button>
  <button
    className="btn+/-"
    onClick={[Function]}
    type="button"
  >
    +/-
  </button>
  <button
    className="btn%"
    onClick={[Function]}
    type="button"
  >
    %
  </button>
  <button
    className="operation-btn"
    onClick={[Function]}
    type="button"
  >
    รท
  </button>
  <button
    className="btn7"
    onClick={[Function]}
    type="button"
  >
    7
  </button>
  <button
    className="btn8"
    onClick={[Function]}
    type="button"
  >
    8
  </button>
  <button
    className="btn9"
    onClick={[Function]}
    type="button"
  >
    9
  </button>
  <button
    className="operation-btn"
    onClick={[Function]}
    type="button"
  >
    x
  </button>
  <button
    className="btn4"
    onClick={[Function]}
    type="button"
  >
    4
  </button>
  <button
    className="btn5"
    onClick={[Function]}
    type="button"
  >
    5
  </button>
  <button
    className="btn6"
    onClick={[Function]}
    type="button"
  >
    6
  </button>
  <button
    className="operation-btn"
    onClick={[Function]}
    type="button"
  >
    -
  </button>
  <button
    className="btn1"
    onClick={[Function]}
    type="button"
  >
    1
  </button>
  <button
    className="btn2"
    onClick={[Function]}
    type="button"
  >
    2
  </button>
  <button
    className="btn3"
    onClick={[Function]}
    type="button"
  >
    3
  </button>
  <button
    className="operation-btn"
    onClick={[Function]}
    type="button"
  >
    +
  </button>
  <button
    className="btn0"
    onClick={[Function]}
    type="button"
  >
    0
  </button>
  <button
    className="btn."
    onClick={[Function]}
    type="button"
  >
    .
  </button>
  <button
    className="operation-btn"
    onClick={[Function]}
    type="button"
  >
    =
  </button>
</div>
`;
